workflow/controller/operator.go

```diff
@@ -490,6 +492,22 @@ func (woc *wfOperationCtx) operate(ctx context.Context) {
 	}
 }
 
+func (woc *wfOperationCtx) releaseLocksForPendingShuttingdownWfs(ctx context.Context) bool {
+	if woc.GetShutdownStrategy().Enabled() && woc.wf.Status.Phase == wfv1.WorkflowPending {
+		// 1. If a workflow is being stopped, all exit handlers should still run.
+		// 2. If a workflow is being terminated, workflow should be marked as succeeded.
+		if woc.GetShutdownStrategy() == wfv1.ShutdownStrategyStop {
+			return false
+		}
+		if woc.controller.syncManager.ReleaseAll(woc.execWf) {
```

Comment:
It seems we treat ShutdownStrategyStop special, whereas the logic would be more correct to treat the Terminate case specially. Because we actually don't care about  ShutdownStrategyStop at all, and should just let it fall down to the `return false` at the end of the function.

```golang
if woc.GetShutdownStrategy().Enabled() && woc.wf.Status.Phase == wfv1.WorkflowPending && woc.GetShutdownStrategy() == wfv1.ShutdownStrategyTerminate {
    if woc.controller.syncManager.ReleaseAll(woc.execWf) {
			woc.log.WithFields(log.Fields{"key": woc.execWf.Name}).Info("Released all locks since this pending workflow is being shutdown")
			woc.markWorkflowSuccess(ctx)
			return true
    }
}
```

Reply:
Oh yes you are right. I should’ve just reverted to my original commit :-) 

Updated.